File: day11/zig/src/main.zig
const std = @import("std");

const Graph = std.StringHashMap(std.ArrayList([]const u8));

const MemoKey = struct {
    node: []const u8,
    has_req1: bool,
    has_req2: bool,
};

const MemoKeyContext = struct {
    pub fn hash(_: @This(), key: MemoKey) u64 {
        var h = std.hash.Wyhash.init(0);
        h.update(key.node);
        h.update(&[_]u8{ @intFromBool(key.has_req1), @intFromBool(key.has_req2) });
        return h.final();
    }
    pub fn eql(_: @This(), a: MemoKey, b: MemoKey) bool {
        return std.mem.eql(u8, a.node, b.node) and a.has_req1 == b.has_req1 and a.has_req2 == b.has_req2;
    }
};

const Memo2Type = std.HashMap(MemoKey, u128, MemoKeyContext, std.hash_map.default_max_load_percentage);

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var stdout_buf: [256]u8 = undefined;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buf);
    const stdout = &stdout_writer.interface;

    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    var graph = Graph.init(allocator);
    defer {
        var iter = graph.valueIterator();
        while (iter.next()) |v| v.deinit(allocator);
        graph.deinit();
    }

    var line_iter = std.mem.splitScalar(u8, content, '\n');
    while (line_iter.next()) |line| {
        const trimmed = std.mem.trim(u8, line, &std.ascii.whitespace);
        if (trimmed.len == 0) continue;

        var parts = std.mem.splitSequence(u8, trimmed, ": ");
        const source = parts.next().?;
        const targets_str = parts.next().?;

        var targets: std.ArrayList([]const u8) = .empty;
        var target_iter = std.mem.splitScalar(u8, targets_str, ' ');
        while (target_iter.next()) |t| {
            try targets.append(allocator, t);
        }

        try graph.put(source, targets);
    }

    var memo1 = std.StringHashMap(u128).init(allocator);
    defer memo1.deinit();
    const part1 = countPaths(&graph, "you", "out", &memo1);

    var memo2 = Memo2Type.init(allocator);
    defer memo2.deinit();

    const part2 = countPathsWithBoth(&graph, "svr", "out", "dac", "fft", false, false, &memo2);

    try stdout.print("Part 1: {d}\n", .{part1});
    try stdout.print("Part 2: {d}\n", .{part2});
    try stdout.flush();
}

fn countPaths(
    graph: *const Graph,
    node: []const u8,
    end: []const u8,
    memo: *std.StringHashMap(u128),
) u128 {
    if (std.mem.eql(u8, node, end)) return 1;

    if (memo.get(node)) |cached| return cached;

    const targets = graph.get(node) orelse return 0;

    var total: u128 = 0;
    for (targets.items) |next| {
        total += countPaths(graph, next, end, memo);
    }

    memo.put(node, total) catch {};
    return total;
}

fn countPathsWithBoth(
    graph: *const Graph,
    node: []const u8,
    end: []const u8,
    req1: []const u8,
    req2: []const u8,
    has_req1: bool,
    has_req2: bool,
    memo: *Memo2Type,
) u128 {
    const now_has_req1 = has_req1 or std.mem.eql(u8, node, req1);
    const now_has_req2 = has_req2 or std.mem.eql(u8, node, req2);

    if (std.mem.eql(u8, node, end)) {
        return if (now_has_req1 and now_has_req2) 1 else 0;
    }

    const key = MemoKey{ .node = node, .has_req1 = now_has_req1, .has_req2 = now_has_req2 };
    if (memo.get(key)) |cached| return cached;

    const targets = graph.get(node) orelse return 0;

    var total: u128 = 0;
    for (targets.items) |next| {
        total += countPathsWithBoth(graph, next, end, req1, req2, now_has_req1, now_has_req2, memo);
    }

    memo.put(key, total) catch {};
    return total;
}

File: day11/rust/src/main.rs
use num_bigint::BigInt;
use std::collections::HashMap;
use std::fs;

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let lines: Vec<&str> = input.lines().filter(|l| !l.is_empty()).collect();
    let graph = build_graph(&lines);

    let path_count_part1 = count_paths(&graph, "you", "out");
    println!("Part 1: {}", path_count_part1);

    let path_count_part2 = count_paths_with_both(&graph, "svr", "out", "dac", "fft");
    println!("Part 2: {}", path_count_part2);
}

fn build_graph(lines: &[&str]) -> HashMap<String, Vec<String>> {
    lines
        .iter()
        .map(|line| {
            let parts: Vec<&str> = line.split(": ").collect();
            let source = parts[0].to_string();
            let targets: Vec<String> = parts[1].split(' ').map(|s| s.to_string()).collect();
            (source, targets)
        })
        .collect()
}

fn count_paths(graph: &HashMap<String, Vec<String>>, start: &str, end: &str) -> BigInt {
    let mut memo: HashMap<String, BigInt> = HashMap::new();

    fn dfs(
        node: &str,
        end: &str,
        graph: &HashMap<String, Vec<String>>,
        memo: &mut HashMap<String, BigInt>,
    ) -> BigInt {
        if node == end {
            return BigInt::from(1);
        }
        if !graph.contains_key(node) {
            return BigInt::from(0);
        }
        if let Some(cached) = memo.get(node) {
            return cached.clone();
        }

        let result: BigInt = graph[node].iter().map(|next| dfs(next, end, graph, memo)).sum();
        memo.insert(node.to_string(), result.clone());
        result
    }

    dfs(start, end, graph, &mut memo)
}

fn count_paths_with_both(
    graph: &HashMap<String, Vec<String>>,
    start: &str,
    end: &str,
    req1: &str,
    req2: &str,
) -> BigInt {
    let mut memo: HashMap<(String, bool, bool), BigInt> = HashMap::new();

    fn dfs(
        node: &str,
        has_req1: bool,
        has_req2: bool,
        end: &str,
        req1: &str,
        req2: &str,
        graph: &HashMap<String, Vec<String>>,
        memo: &mut HashMap<(String, bool, bool), BigInt>,
    ) -> BigInt {
        let now_has_req1 = has_req1 || node == req1;
        let now_has_req2 = has_req2 || node == req2;

        if node == end {
            return if now_has_req1 && now_has_req2 {
                BigInt::from(1)
            } else {
                BigInt::from(0)
            };
        }
        if !graph.contains_key(node) {
            return BigInt::from(0);
        }

        let key = (node.to_string(), now_has_req1, now_has_req2);
        if let Some(cached) = memo.get(&key) {
            return cached.clone();
        }

        let result: BigInt = graph[node]
            .iter()
            .map(|next| dfs(next, now_has_req1, now_has_req2, end, req1, req2, graph, memo))
            .sum();
        memo.insert(key, result.clone());
        result
    }

    dfs(start, false, false, end, req1, req2, graph, &mut memo)
}
